lib/src/native/realm_core.dart

```diff
@@ -2907,45 +2907,63 @@ extension on Pointer<realm_value_t> {
     if (this == nullptr) {
       throw RealmException("Can not convert nullptr realm_value to Dart value");
     }
+    return ref.toDartValueByRef(realm);
+  }
+}
 
-    switch (ref.type) {
+extension on realm_value_t {
+  Object? toDartValueByRef(Realm? realm) {
+    switch (type) {
       case realm_value_type.RLM_TYPE_NULL:
         return null;
       case realm_value_type.RLM_TYPE_INT:
-        return ref.values.integer;
+        return values.integer;
       case realm_value_type.RLM_TYPE_BOOL:
-        return ref.values.boolean;
+        return values.boolean;
       case realm_value_type.RLM_TYPE_STRING:
-        return ref.values.string.data.cast<Utf8>().toRealmDartString(length: ref.values.string.size)!;
+        return values.string.data.cast<Utf8>().toRealmDartString(length: values.string.size)!;
       case realm_value_type.RLM_TYPE_FLOAT:
-        return ref.values.fnum;
+        return values.fnum;
       case realm_value_type.RLM_TYPE_DOUBLE:
-        return ref.values.dnum;
+        return values.dnum;
       case realm_value_type.RLM_TYPE_LINK:
-        final objectKey = ref.values.link.target;
-        final classKey = ref.values.link.target_table;
+        if (realm == null) {
+          return null;
+        }
+        final objectKey = values.link.target;
+        final classKey = values.link.target_table;
         if (realm.metadata.getByClassKeyIfExists(classKey) == null) return null; // temprorary workaround to avoid crash on assertion
         return realmCore._getObject(realm, classKey, objectKey);
       case realm_value_type.RLM_TYPE_BINARY:
         throw Exception("Not implemented");
       case realm_value_type.RLM_TYPE_TIMESTAMP:
-        final seconds = ref.values.timestamp.seconds;
-        final nanoseconds = ref.values.timestamp.nanoseconds;
+        final seconds = values.timestamp.seconds;
+        final nanoseconds = values.timestamp.nanoseconds;
         return DateTime.fromMicrosecondsSinceEpoch(seconds * _microsecondsPerSecond + nanoseconds ~/ _nanosecondsPerMicrosecond, isUtc: true);
       case realm_value_type.RLM_TYPE_DECIMAL128:
-        var decimal = ref.values.decimal128; // NOTE: Does not copy the struct!
+        var decimal = values.decimal128; // NOTE: Does not copy the struct!
         decimal = _realmLib.realm_dart_decimal128_copy(decimal); // This is a workaround to that
         return Decimal128Internal.fromNative(decimal);
       case realm_value_type.RLM_TYPE_OBJECT_ID:
-        return ObjectId.fromBytes(cast<Uint8>().asTypedList(12));
+        return ObjectId.fromBytes(values.object_id.bytes.toIntList(ObjectId.byteLength));
```

Comment:
the hardcoded value here is removed but not on the `case realm_value_type.RLM_TYPE_UUID`
Either we hard code as it was or we make sure we are passing this information from native side and we change the `RLM_TYPE_UUID` case as well since it has `kUuidBytes` constant as well.